core/src/main/scala/nl/vroste/zio/kinesis/client/Producer.scala

```diff
@@ -104,14 +110,53 @@ final case class ProducerSettings(
   failedDelay: Duration = 100.millis,
   metricsInterval: Duration = 30.seconds,
   updateShardInterval: Duration = 30.seconds,
-  aggregate: Boolean = false,
   allowedErrorRate: Double = 0.05,
-  shardPredictionParallelism: Int = 8
+  shardPrediction: Producer.ShardPrediction = Producer.ShardPrediction.Enabled(8),
```

Comment:
My goal for structuring configuration in this way is to have as few illegal setting combinations / unused settings as possible (aggregation requiring shardprediction is still a bit ugly though).

Added another setting to toggle throttling.

Reply:
That's a good goal in general. But in this case I keep thinking that it would be better to have the settings that control the more top-level features as their own parameters: aggregation and throttling. Currently throttling is a sub-pararameter of shard prediction, which is not a feature in itself but just a shared implementation detail of the throttling and aggregation features. 

IMO it's fine to have a parameter `shardPredictionParallelism` with a comment like "this setting is only used when throttling or aggregation is enabled".

Reply:
Adjusted, what do you think about the current version?

I agree with your perspective about throttling. Making it a top-level setting makes sense, even if it requires prediction to be enabled. I would still prefer to keep shardPredictionParallelism a subsetting of shardprediction though, as it really has no meaning on its own.